install_files/ansible-base/roles/ossec/files/process_submissions_today.sh

```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
+
+function send_encrypted_alarm() {
+    /var/ossec/send_encrypted_alarm.sh "$1"
+}
+
+function main() {
+    local sender
+    local stdin
+    sender=${1:-send_encrypted_alarm}
+    stdin="$(< /dev/stdin)"
+
+    local count
+    count=$(echo "$stdin" | perl -ne 'print scalar(<>) and exit if(/ossec: output/);')
+    if [[ "$count" =~ ^[0-9]+$ ]] ; then
+        export SUBJECT="Submissions in the past 24h"
+        #
+        # whitespaces below are so the size of both messages are exactly the same
+        #
+        if [[ "$count" -gt "0" ]] ; then
+            echo "There has been submission activity in the past 24 hours.   "
+            echo "You should login and check SecureDrop. "
+        else
+            echo "There has been no submission activity in the past 24 hours."
+            echo "You do not need to login to SecureDrop."
+        fi | $sender journalist
+    else
+        export SUBJECT="SecureDrop Submissions Error"
+        (
+            echo "$0 failed to find the number of submissions in the following OSSEC alert"
+            echo
+            echo "$stdin"
+        ) | $sender ossec
+    fi
+}
```

Comment:
Since OSSEC by default [groups alerts](https://github.com/ossec/ossec-hids/blob/72641d6f22c63b97f290ae22d47a79032b56d0fd/etc/internal_options.conf#L59) together, it's possible to get a "SecureDrop Submissions Error" sent to the administrator if there are multiple alerts that occur in a short time span. This alert contains both the contents of the submissions text file on disk along with the other alert(s) that occur. One way to reproducibly trigger this is by rebooting the VMs:

```
/var/ossec/process_submissions_today.sh failed to find the number of submissions in the following OSSEC alert

>From ossec@freedom.press  Sat Feb 24 18:44:02 2018
Return-Path: <ossec@freedom.press>
X-Original-To: root@localhost
Delivered-To: root@localhost
Received: from notify.ossec.net (mon-staging [127.0.0.1])
	by ossec.server (Postfix) with SMTP id 7170C1202F6
	for <root@localhost>; Sat, 24 Feb 2018 18:44:02 +0000 (UTC)
To: <root@localhost>
From: OSSEC HIDS <ossec@freedom.press>
Date: Sat, 24 Feb 2018 18:44:02 +0000
Subject: OSSEC Notification - (app-staging) 10.0.1.2 - Alert level 7
Message-Id: <20180224184402.7170C1202F6@ossec.server>


OSSEC HIDS Notification.
2018 Feb 24 18:43:50

Received From: (app-staging) 10.0.1.2->netstat -tan |grep LISTEN |grep -v 127.0.0.1 | sort
Rule: 533 fired (level 7) -> "Listened ports status (netstat) changed (new port opened or closed)."
Portion of the log(s):

ossec: output: 'netstat -tan |grep LISTEN |grep -v 127.0.0.1 | sort':
tcp        0      0 0.0.0.0:111             0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:55436           0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:6001            0.0.0.0:*               LISTEN     
tcp6       0      0 :::111                  :::*                    LISTEN     
tcp6       0      0 :::40520                :::*                    LISTEN     
tcp6       0      0 :::6001                 :::*                    LISTEN     
Previous output:
ossec: output: 'netstat -tan |grep LISTEN |grep -v 127.0.0.1 | sort':
tcp        0      0 0.0.0.0:111             0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:53212           0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:6001            0.0.0.0:*               LISTEN     
tcp6       0      0 :::111                  :::*                    LISTEN     
tcp6       0      0 :::54287                :::*                    LISTEN     
tcp6       0      0 :::6001                 :::*                    LISTEN     



 --END OF NOTIFICATION



OSSEC HIDS Notification.
2018 Feb 24 18:43:50

Received From: (app-staging) 10.0.1.2->head -1 /var/lib/securedrop/submissions_today.txt | grep '^[0-9]*$'
Rule: 400600 fired (level 7) -> "Number of submissions received in the past 24h."
Portion of the log(s):

ossec: output: 'head -1 /var/lib/securedrop/submissions_today.txt | grep '^[0-9]*$'':
1



 --END OF NOTIFICATION
```

Reply:
**very** good catch ! Will add [do_not_group](https://ossec-docs.readthedocs.io/en/latest/syntax/head_ossec_config.email_alerts.html#element-do_not_group) to prevent that

Reply:
do_not_group does not actually help because the alerts are grouped and do_not_group applies to mails not being grouped which is a different problem. I'm now trying a different approach which is to lower the level of the alert (level 1) so it is not grouped with other level 7 alerts. And I'm creating an ossec.conf email_alerts dedicated to this rule. It may not work as I think it does though...

Reply:
I made it so two alerts are sent at the same time and got two separate emails. Although this is not documented, I think it makes sense that the level 1 alert regarding submissions is not grouped with level 7 alerts. @redshiftzero it would be nice if you could repeat what you did to see the problem and verify if it really is gone. If you can document the steps I'd be happy to do it myself. I tried to reboot the staging vm but I doubt that's what you did because it requires manually running the ossec server and we're no longer in the same situation as the one you experienced ;-)

<pre>
# mailq
-Queue ID- --Size-- ----Arrival Time---- -Sender/Recipient-------
17FC41403F4     1105 Sun Feb 25 14:18:29  ossec@ossec.server
(Host or domain name not found. Name service error for name=smtp.faketld type=A: Host not found)
                                         journalist@ossec.test

15D251403F6     1512 Sun Feb 25 14:18:29  ossec@ossec.server
(Host or domain name not found. Name service error for name=smtp.faketld type=A: Host not found)
                                         ossec@ossec.test

-- 2 Kbytes in 2 Requests.
# postcat -q 15D251403F6 | gpg --homedir /var/ossec/.gnupg/ --decrypt                                           
OSSEC HIDS Notification.
2018 Feb 25 14:18:27

Received From: (app-staging) 10.0.1.2->syscheck
Rule: 550 fired (level 7) -> "Integrity checksum changed."
Portion of the log(s):

Integrity checksum changed for: '/var/ossec/etc/ossec.conf'
Size changed from '4608' to '4657'
Old md5sum was: 'b998b945f355a8d37883e124960b0214'
New md5sum is : '124e883d6e69498ae3a14a697332afee'
Old sha1sum was: '52f8d74b7087eb69efaae51921d8b47714855656'
New sha1sum is : '22e0dfca73d67879ca89e815ef9530b4784ae638'



 --END OF NOTIFICATION

# postcat -q 17FC41403F4 | gpg --homedir /var/ossec/.gnupg/ --decrypt                                           

There has been no submission activity in the past 24 hours.
You do not need to login to SecureDrop.
</pre>

Reply:
Nice fix - with this change the submissions error message does not occur :clap:.

For posterity, I verified this by:

0. checking out the latest changes here
1. `make build-debs`
2. `vagrant provision /staging/`
3. `vagrant reload /staging/` - this is what produced the error originally
4. Checking my OSSEC alerts

Reply:
That makes me curious because ... I don't have OSSEC or postfix running when I provision staging? Did you tweak the vagrant files ? Or am I misisng something ?

Reply:
Ah so when I'm testing a PR where it would be convenient to have both speedy SSH _and_ OSSEC emails, I edit the staging group variables like so:

```
diff --git a/install_files/ansible-base/group_vars/staging.yml b/install_files/ansible-base/group_vars/staging.yml
index c27e5a96..66612ab5 100644
--- a/install_files/ansible-base/group_vars/staging.yml
+++ b/install_files/ansible-base/group_vars/staging.yml
@@ -31,21 +31,10 @@ apache_logging_level: "info"
 ### Used by the mon role ###
 # The OSSEC alert GPG public key has to be in the install_files/ansible-base/ or
 # install_files/ansible-base/roles/app/files/ directory
-ossec_alert_gpg_public_key: "test_admin_key.pub"
-ossec_gpg_fpr: "53E1113AC1F25027BA5D475B1141E2BBB5E53711"
-ossec_alert_email: "ossec@ossec.test"
-journalist_alert_gpg_public_key: "test_admin_key.pub"
-journalist_gpg_fpr: "53E1113AC1F25027BA5D475B1141E2BBB5E53711"
-journalist_alert_email: "journalist@ossec.test"
-smtp_relay: "smtp.faketld"
-smtp_relay_port: "587"
-sasl_username: "test"
-sasl_domain: "ossec.test"
-sasl_password: "password123"
 
 # Disable Postfix in staging, so we don't hammer Google mail relays
 # with known-bad credentials.
-postfix_enable_service: no
+postfix_enable_service: yes
```

Reply:
nice trick !